pkg/util/limiter/query_limiter.go

```diff
@@ -0,0 +1,78 @@
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+type queryLimiterCtxKey struct{}
+
+var (
+	qlCtxKey        = &queryLimiterCtxKey{}
+	errMaxSeriesHit = "The query hit the max number of series limit while fetching chunks (limit: %d)"
+)
+
+type QueryLimiter struct {
+	uniqueSeriesMx sync.RWMutex
+	uniqueSeries   map[model.Fingerprint]struct{}
+
+	maxSeriesPerQuery int
+}
+
+// NewQueryLimiter makes a new per-query limiter. Each query limiter
+// is configured using the `maxSeriesPerQuery` limit.
+func NewQueryLimiter(maxSeriesPerQuery int) *QueryLimiter {
+	return &QueryLimiter{
+		uniqueSeriesMx: sync.RWMutex{},
+		uniqueSeries:   map[model.Fingerprint]struct{}{},
+
+		maxSeriesPerQuery: maxSeriesPerQuery,
+	}
+}
+
+func AddQueryLimiterToContext(ctx context.Context, limiter *QueryLimiter) context.Context {
+	return context.WithValue(ctx, qlCtxKey, limiter)
+}
+
+// QueryLimiterFromContextWithFallback returns a QueryLimiter from the current context.
+// If there is not a QueryLimiter on the context it will return a new no-op limiter.
+func QueryLimiterFromContextWithFallback(ctx context.Context) *QueryLimiter {
+	ql, ok := ctx.Value(qlCtxKey).(*QueryLimiter)
+	if !ok {
+		// If there's no limiter return a new unlimited limiter as a fallback
+		ql = NewQueryLimiter(0)
+	}
+	return ql
+}
+
+// AddSeries adds the input series and returns an error if the limit is reached.
+func (ql *QueryLimiter) AddSeries(seriesLabels []cortexpb.LabelAdapter) error {
+	// If the max series is unlimited just return without managing map
+	if ql.maxSeriesPerQuery == 0 {
+		return nil
+	}
+
+	ql.uniqueSeriesMx.Lock()
+	defer ql.uniqueSeriesMx.Unlock()
+
+	ql.uniqueSeries[client.FastFingerprint(seriesLabels)] = struct{}{}
```

Comment:
Nit: To minimize time during which the lock is held, it is preferable to do computations that don't need the lock before locking. In this case it's call to `client.FastFingerprint`. It's unlikely to be a big deal here, but still a good practice.

I also find it little strange that we always add to the map before checking the limit. If entries in the map were larger, it could lead to unexpected memory problems. Here it's not a problem.

Reply:
> I also find it little strange that we always add to the map before checking the limit.  

I figured since we don't know if the series has a unique fingerprint and will put us over the limit, we can't check until after we've added/not added the passed in series fingerprint